accounting/bank/fields.py

```diff
@@ -0,0 +1,51 @@
+# -*- coding: utf-8 -*-
+#
+# OpenCraft -- tools to aid developing and hosting free software projects
+# Copyright (C) 2015-2017 OpenCraft <contact@opencraft.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+"""
+Bank fields used by any of the Bank models.
+"""
+
+from django.db import models
+from django.utils.translation import ugettext_lazy as _
+
+
+class AUBankStateBranchField(models.CharField):
+    """
+    A model field that checks that the value is a valid Australian Bank State Branch (BSB) code.
+
+    BSBs are 6-digit codes. From https://www.thebsbnumbers.com/:
+        - The First two digits (XX) specify the parent financial institution.
+        - Third digit (Y) specifies the state where the branch is located.
+        - Fourth, fifth and sixth digits (ZZZ) specify the branch location.
+    """
```

Comment:
As mentioned above and the email thread on all@, I'd like to avoid handling specific use cases for the different countries manually in our code. Can we delegate that part to Transferwise, only collecting the information they ask for through their API, and having one approach that works for all countries? Otherwise we'll need to update the code evey time their is a change in legislation, or whenever someone joins from a country with a set of rules we haven't implemented yet.

Reply:
So there probably isn't going to be a significant enough change in legislation often enough to have that require code changes, and I've designed this so that if we do have to add a special 'code' or something for a new country, it should be trivial to do so (as trivial as just adding the new field to the model; i.e. 20~ min of work). But I don't think even *that* would be required, since I got  these requirements from TransferWise already.

I think the issue we have is I envisioned that recipient accounts would automatically be created through this app's automation, whereas you envisioned that they already would be created and just re-used through the app. This is where I'm confused, and I think our discussions aren't helping my confusion necessarily: where does the automation start and where does it end? What are its boundaries? I feel like we're only figuring this out just now.

Reply:
I'm going to adopt Braden's recommendation and use a JSONField to hold these attributes.

Reply:
👍